### PR TITLE
Fix constraints without any refs

### LINKED ISSUE #5530: ISE migrating expression constraint (true)
- EdgeDB Version: 3.0-rc.1+5979043
- EdgeDB CLI Version: 3.0.0+8b024db
- OS Version: ubuntu 22.04

Steps to Reproduce:
1. put the schema in `default.esdl`
2. 
```bash
$ edgedb migration create
Created /home/fmoor/src/sample/dbschema/migrations/00001.edgeql, id: m1ok5bhgpjxqxiihqqibbzyddu5t3xtipj34xiio7xhpqd5w6vdmha

$ edgedb migration apply
edgedb error: InternalServerError: relation "edgedbpub.d1e7c309-f8ed-11ed-ab0c-d5c6508df419" does not exist
  Hint: This is most likely a bug in EdgeDB. Please consider opening an issue ticket at https://github.com/edgedb/edgedb/issues/new?template=bug_report.md
  Server traceback:
      edb.errors.InternalServerError: relation "edgedbpub.d1e7c309-f8ed-11ed-ab0c-d5c6508df419" does not exist
edgedb error: error in one of the migrations

```

Schema:
```esdl
module default {
  type Log {
    target_name -> str {
      constraint expression on (
        true
      );
    };
  }
}
```

### PR DESCRIPTION
**Pull request by aljazerzen**
Closes #5530

Vast majority of changes here are:
- type annotations for pgsql/schemamech.py and friends,
- refactors schemamech.ConstraintMech into plain functions,
- refactor many `Dict[str, Any]` into data classes (which are better with type checking and will be fasterrr in Python 11)

### CODE REVIEW

File: edb/pgsql/schemamech.py [hunk 1]
```diff
@@ -50,361 +53,412 @@
...
+    elif ref_tables:
+        subject_db_name, info = next(iter(ref_tables.items()))
+        table_type = info[0][3].table_type
+    else:
+        # the expression does don't have any refs: default to the subject table
+
+        if isinstance(subject, s_pointers.Pointer):
+            subject_table = subject.get_source(schema)
+        else:
+            subject_table = subject
```
Comment (aljazerzen) [inline edb/pgsql/schemamech.py:287]: This is the only impactful change. Should I put it into a separate PR?

Comment (elprans) [reply edb/pgsql/schemamech.py:287]: (or just make sure it's a separate commit)

Comment (aljazerzen) [reply edb/pgsql/schemamech.py:287]: I've squashed it into two commits that will be "rebase and merge"d.

Comment (elprans) [reply edb/pgsql/schemamech.py:287]: Hm, except this wouldn't cherry-pick cleanly without the refactor?  How comfortable are we with cherry-picking the refactor?

**Review by msullivan — APPROVED** [review]
This is a lot better
